src/app/ClusterObjectUtils.h

```diff
@@ -0,0 +1,312 @@
+}
+
+/*
+ * @brief
+ *
+ * This specific variant that encodes cluster objects (like structs, commands, events) to TLV
+ * depends on the presence of an Encode method on the object to present. The signature of that method
+ * is as follows:
+ *
+ * CHIP_ERROR <Object>::Encode(TLVWriter &writer, uint64_t tag);
+ *
+ *
+ */
+template <typename X, typename std::enable_if<std::is_class<X>::value, X>::type* = nullptr>
+CHIP_ERROR Encode(TLV::TLVWriter &writer, uint64_t tag, X &x) {
+    return x.Encode(writer, tag);
+}
+
+//
+// Decode
+//
+template <typename X, typename std::enable_if<std::is_integral<X>::value, int>::type = 0>
+CHIP_ERROR Decode(TLV::TLVReader &reader, X &x) {
+    return reader.Get(x);
+}
+
+CHIP_ERROR Decode(TLV::TLVReader &reader, chip::ByteSpan &x) {
+    VerifyOrReturnError(reader.GetType() == TLV::kTLVType_ByteString, CHIP_ERROR_UNEXPECTED_TLV_ELEMENT);
+    return reader.Get(x);
+}
+
+CHIP_ERROR Decode(TLV::TLVReader &reader, chip::Span<char> &x) {
+    chip::ByteSpan bs;
+  
+    //
+    // Like for byte string, a span is returned that points into the underlying
+    // buffer in the reader itself.
+    //
+    VerifyOrReturnError(reader.GetType() == TLV::kTLVType_UTF8String, CHIP_ERROR_UNEXPECTED_TLV_ELEMENT); 
+    ReturnErrorOnFailure(reader.Get(bs));
+    x = chip::Span<char>((char *)bs.data(), (size_t)bs.size());
+    return CHIP_NO_ERROR;
+}
+
+/*
+ * @brief
+ *
+ * This specific variant that decodes cluster objects (like structs, commands, events) from TLV
+ * depends on the presence of a Decode method on the object to present. The signature of that method
+ * is as follows:
+ *
+ * CHIP_ERROR <Object>::Decode(TLVReader &reader);
+ *
+ */
+template <typename X, typename std::enable_if<std::is_class<X>::value, X>::type* = nullptr>
+CHIP_ERROR Decode(TLV::TLVReader &reader, X &x) {
+    return x.Decode(reader);
+}
+
+/*
+ * @brief
+ *
+ * This class provides an iteratable decoder of list items within TLV payloads
+ * such that no memory has to be provided ahead of time to store the entirety of the decoded
+ * list contents.
+ *
+ * Specifically, this provides a read-only input iterator that overloads the '->' operator to dynamically
+ * decode the list item at a given position in the stored TLVReader and return it to the caller.
+ *
+ */
+template <typename T>
+class IteratableList {
+public:
+    void SetReader(TLV::TLVReader& reader) { mReader = reader; }
+
+    size_t size() {
+        CHIP_ERROR err;
+        TLV::TLVReader reader;
+        reader.Init(mReader);
+        size_t count = 0;
+
+        while ((err = reader.Next()) == CHIP_NO_ERROR) {
+            count++; 
+        }
+
+        return count;
+    }
+
+    class Iterator {
+    public:
+        /*
+         * @brief
+         *
+         * Since neither the iterator constructor nor the operator overload permit sending errors back,
+         * encapsulate a tuple of the value and error that is then returned to the caller when decoding
```

Comment:
I would much prefer it if we had APIs that more straightforwardly returned errors.  For example, a `Next` method instead of `operator->`.  That would avoid the need for this extra struct type, I would think.

Reply:
Alternatively, instead of rolling our own, can optional<T> be used or extended?  I've seen that approach in some other parts of the sdk, and that would prevent people having to learn the special semantics of this struct, which does essentially the same thing, but with fewer features.

I assume you could still extend that to store a CHIP_ERROR too.

Reply:
Ah, and I see that we did implement most aspects of optional<T>, including the -> operator.

Seems like further reason to just extend optional<T> with addition of CHIP_ERROR.  Is the justification that that leads to a possible mismatch between status and hasValue?

Reply:
I've since pivoted to a model that reduces the use of overloads, and just creates a single, `Next` method that increments and decodes at the same time. Please check it out!